src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs

```diff
@@ -1745,9 +1758,33 @@ protected void AfterMembersChecks(BindingDiagnosticBag diagnostics)
                 }
             }
 
-            if (IsFileLocal && (object?)ContainingType != null)
+            if (AssociatedFileIdentifier is { } fileIdentifier)
             {
-                diagnostics.Add(ErrorCode.ERR_FileTypeNested, location, this);
+                Debug.Assert(IsFileLocal);
+
+                // A well-behaved file-local type only has declarations in one syntax tree.
+                // There may be multiple syntax trees across declarations in error scenarios,
+                // but we're not interested in handling that for the purposes of producing this error.
+                var tree = declaration.Declarations[0].SyntaxReference.SyntaxTree;
+                if (fileIdentifier.FilePathChecksumOpt.IsDefault)
+                {
+                    Debug.Assert(fileIdentifier.EncoderFallbackErrorMessage is not null);
```

Comment:
Minor: Would it make sense to check for the error message rather than the check sum? `if (fileIdentifier.EncoderFallbackErrorMessage is { } message) { ... }`